SCons/Tool/vcpkg.py

```diff
@@ -0,0 +1,613 @@
+
+def _vcpkg_print(verbosity, *args):
+    """If the user wants to see messages of 'verbosity', prints *args with a 'vcpkg' prefix"""
+
+    if verbosity <= _max_verbosity:
+        print('vcpkg: ', end='')
+        print(*args)
+
+
+def _get_built_vcpkg_full_version(vcpkg_exe):
+    """Runs 'vcpkg version' and parses the version string from the output"""
+
+    if not vcpkg_exe.exists():
+        raise InternalError(vcpkg_exe.get_path() + ' does not exist')
+
+    line = subprocess.check_output([vcpkg_exe.get_abspath(), 'version'], text = True)
+    match_version = re.search(r' version (\S+)', line)
+    if (match_version):
+        version = match_version.group(1)
+        _vcpkg_print(Debug, 'vcpkg full version is "' + version + '"')
+        return version
+    raise InternalError(vcpkg_exe.get_path() + ': failed to parse version string')
+
+
+def _get_built_vcpkg_base_version(vcpkg_exe):
+    """Returns just the base version from 'vcpkg version' (i.e., with any '-whatever' suffix stripped off"""
+    full_version = _get_built_vcpkg_full_version(vcpkg_exe)
+
+    # Allow either 2022-02-02 or 2022.02.02 date syntax
+    match = re.match(r'^(\d{4}[.-]\d{2}[.-]\d{2})', full_version)
+    if match:
+        base_version = match.group(1)
+        _vcpkg_print(Debug, 'vcpkg base version is "' + base_version + '"')
+        return base_version
+
+    _vcpkg_print(Debug, 'vcpkg base version is identical to full version "' + full_version + '"')
+    return full_version
+
+
+def _get_source_vcpkg_version(vcpkg_root):
+    """Parses the vcpkg source code version out of VERSION.txt"""
+
+    if not vcpkg_root.exists():
+        raise InternalError(vcpkg_root.get_path() + ' does not exist')
+
+    # Older vcpkg versions had the source version in VERSION.txt
+    version_txt = vcpkg_root.File('toolsrc/VERSION.txt')
+    if version_txt.exists():
+        _vcpkg_print(Debug, "Looking for source version in " + version_txt.get_path())
+        for line in open(version_txt.get_abspath()):
+            match_version = re.match(r'^"(.*)"$', line)
+            if (match_version):
+                version = match_version.group(1)
+
+                # Newer versions of vcpkg have a hard-coded invalid date in the VERSION.txt for local builds,
+                # and the actual version comes from the bootstrap.ps1 script
+                if version != '9999.99.99':
+                    _vcpkg_print(Debug, 'Found valid vcpkg source version "' + version + '" in VERSION.txt')
+                    return version
+                else:
+                    _vcpkg_print(Debug, 'VERSION.txt contains invalid vcpkg source version "' + version + '"')
+                    break
+
+    # Newer versions of bootstrap-vcpkg simply download a pre-built executable from GitHub, and the version to download
+    # is hard-coded in a deployment script.
+    bootstrap_ps1 = vcpkg_root.File('scripts/bootstrap.ps1')
```

Comment:
is this a win32 only implementation?

Reply:
Not Win32-only. Notice that I'm not _executing_ the powershell script, just scraping the version string out of it. I may try to get vcpkg to expose this in a simple text file in the future.